File: library/src/main/java/no/wtw/android/architectureutils/adapter/ListItemAdapter.java
package no.wtw.android.architectureutils.adapter;

import android.content.Context;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;

import org.androidannotations.annotations.EBean;
import org.androidannotations.annotations.RootContext;

import androidx.annotation.DrawableRes;
import no.wtw.android.architectureutils.R;
import no.wtw.android.architectureutils.model.Listable;
import no.wtw.android.architectureutils.view.ListItemView;
import no.wtw.android.architectureutils.view.ListItemView_;

@EBean
public class ListItemAdapter extends RecyclerViewAdapterBase<Listable, ListItemView> {

    @RootContext
    protected Context context;

    private OnItemClickListener<Listable, ListItemView> extraClickListener;
    @DrawableRes
    private int extraButtonClickResource;

    @Override
    protected ListItemView onCreateItemView(ViewGroup parent, int viewType) {
        return ListItemView_.build(context);
    }

    @Override
    public void onBindViewHolder(ViewWrapper<Listable, ListItemView> viewHolder, int position) {
        super.onBindViewHolder(viewHolder, position);
        ImageButton extraClickButton = viewHolder.getView().findViewById(R.id.extra_button);
        extraClickButton.setOnClickListener(v -> onItemExtraClick(viewHolder.getAdapterPosition(), viewHolder.getView(), filteredItems.get(position)));
        extraClickButton.setVisibility(extraClickListener == null ? View.GONE : View.VISIBLE);
        extraClickButton.setImageResource(extraButtonClickResource);
    }

    public void setonItemExtraClickListener(OnItemClickListener<Listable, ListItemView> listener, @DrawableRes int iconResourceId) {
        this.extraClickListener = listener;
        this.extraButtonClickResource = iconResourceId;
        notifyDataSetChanged();
    }

    public void onItemExtraClick(int position, ListItemView view, Listable data) {
        if (extraClickListener != null)
            extraClickListener.onItemClick(position, view, data);
    }

}

File: library/src/main/java/no/wtw/android/architectureutils/adapter/ListItemAdapter.kt
package no.wtw.android.architectureutils.adapter

import android.content.Context
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import androidx.annotation.DrawableRes
import no.wtw.android.architectureutils.R
import no.wtw.android.architectureutils.model.Listable
import no.wtw.android.architectureutils.view.ListItemView

open class ListItemAdapter(
    private val context: Context,
) : RecyclerViewAdapterBase<Listable, ListItemView>() {

    private var extraClickListener: OnItemClickListener<Listable, ListItemView>? = null

    @DrawableRes
    private var extraButtonClickResource = 0

    override fun onCreateItemView(parent: ViewGroup, viewType: Int) =
        ListItemView(context)

    override fun onBindViewHolder(viewHolder: ViewWrapper<Listable, ListItemView>, position: Int) {
        super.onBindViewHolder(viewHolder, position)
        val extraClickButton = viewHolder.view.findViewById<ImageButton>(R.id.extra_button)
        extraClickButton.setOnClickListener { onItemExtraClick(viewHolder.bindingAdapterPosition, viewHolder.view, filteredItems[position]) }
        extraClickButton.visibility = if (extraClickListener == null) View.GONE else View.VISIBLE
        extraClickButton.setImageResource(extraButtonClickResource)
    }

    fun setonItemExtraClickListener(listener: OnItemClickListener<Listable, ListItemView>?, @DrawableRes iconResourceId: Int) {
        extraClickListener = listener
        extraButtonClickResource = iconResourceId
        notifyDataSetChanged()
    }

    fun onItemExtraClick(position: Int, view: ListItemView, data: Listable) {
        extraClickListener?.onItemClick(position, view, data)
    }

}